### ISSUE TITLE
Can we do waypoint training on Macbook Pro with MPS GPU?

### ISSUE THREAD

**Issue by Beethoven-Q** [priyasundaresan/sphinx#4]
Hi! Thank you so much for this elegant open source repo, which is pretty educational. 
Since you provided a `mac_env.yml` for the mac environment setup, I think this project should be mac-friendly. 
I am trying to run both the waypoint training and evaluation on my Macbook pro, with an M2 MAX chip which I think is powerful enough. 
I changed all `.cuda()` in the code to `.to(device)` and define device to be:
`if torch.cuda.is_available():
    device = torch.device("cuda")
elif torch.backends.mps.is_available():
    device = torch.device("mps")
else:
    device = torch.device("cpu")`

And also changed `torch.cuda.synchronize()`
to:
`if device.type == "cuda":
    torch.cuda.synchronize()` in train_waypoint.py

Also changed `with torch.backends.cuda.sdp_kernel(enable_math=False):
    attn_v = torch.nn.functional.scaled_dot_product_attention(...)`
to 
`if x.is_cuda and torch.cuda.is_available():
    # CUDA device => can use sdp_kernel
    with torch.backends.cuda.sdp_kernel(enable_math=False):
        attn_v = torch.nn.functional.scaled_dot_product_attention(
            q, k, v, dropout_p=0.0, attn_mask=attn_mask
        )
else:
    # fallback for CPU or MPS
    attn_v = torch.nn.functional.scaled_dot_product_attention(
        q, k, v, dropout_p=0.0, attn_mask=attn_mask
    )` 
in train_waypoint.py

Now the `eval_sim_waypoint.py` runs pretty well on my macbook! Even though my waypoint model was trained on NVDA CUDA, it can also be evaluated on my mac with rendering on to show its performance. 

But when running the `train_waypoint.py`, I met some issues. The program can actually be started, a wandb running is created and the terminal log appears nice. But after a well it terminated with a bus error: 
`[1]    16267 bus error  python scripts/train_waypoint.py --config_path cfgs/waypoint/can.yaml`

**The complete terminal output is like below:**
(sphinx_env) ➜  sphinx git:(main) ✗ python scripts/train_waypoint.py --config_path cfgs/waypoint/can.yaml
Running on device: mps
=================config=================
batch_size: 64
cosine_schedule: 1
dataset:
  aug_interpolate: 0.2
  aug_rotate: 0
  aug_translate: 1
  fps: 0
  is_real: 0
  path: data/can
  radius: 0.05
  repeat: 20
  split_percent: 0.85
  split_seed: 1
  use_dist: 1
epoch: 100
eval_per_epoch: 10
grad_clip: 1.0
lr: 0.0001
num_eval_episode: 100
num_workers: 2
save_dir: exps/waypoint/tsfm1
save_per: -1
seed: 1
train_split: all
use_ema: 1
use_wb: 1
waypoint:
  custom_init: 0
  drop: 0.1
  embed_dim: 512
  final_ln: 1
  npoints: 1024
  num_head: 8
  num_layer: 6
  per_point_rot: 0
  pred_off: 1
  pred_point: 1
  preset: small
  topk_eval: 3
  topk_train: 0
  use_euler: 1
wandb: Currently logged in as: <xxxxx>. Use `wandb login --relogin` to force relogin
wandb: wandb version 0.19.4 is available!  To upgrade, please run:
wandb:  $ pip install wandb --upgrade
wandb: Tracking run with wandb version 0.17.5
wandb: Run data is saved locally in <xxxxxxxxxxxxx>
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run tsfm1
wandb:  View project at <xxxxxxxxxxxxxx>
wandb:  View run at <xxxxxxxxxxxxxxxxxx>
Creating all dataset with 20 demos
[robosuite WARNING] No private macro file found! (macros.py:53)
[robosuite WARNING] It is recommended to use a private macro file (macros.py:54)
[robosuite WARNING] To setup, run: python /Users/xxxxxxx/anaconda3/envs/sphinx_env/lib/python3.9/site-packages/robosuite/scripts/setup_macros.py (macros.py:55)
objc[16911]: Class GLFWApplicationDelegate is implemented in both /Users/xxxxxxxx/anaconda3/envs/sphinx_env/lib/python3.9/site-packages/open3d/cpu/pybind.cpython-39-darwin.so (0x31d47ee20) and /Users/xxxxxxxxx/anaconda3/envs/sphinx_env/lib/python3.9/site-packages/glfw/libglfw.3.dylib (0x339d45bd0). One of the two will be used. Which one is undefined.
objc[16911]: Class GLFWWindowDelegate is implemented in both /Users/xxxxxxxxx/anaconda3/envs/sphinx_env/lib/python3.9/site-packages/open3d/cpu/pybind.cpython-39-darwin.so (0x31d47eda8) and /Users/xxxxxxxxx/anaconda3/envs/sphinx_env/lib/python3.9/site-packages/glfw/libglfw.3.dylib (0x339d45bf8). One of the two will be used. Which one is undefined.
objc[16911]: Class GLFWContentView is implemented in both /Users/xxxxxxxx/anaconda3/envs/sphinx_env/lib/python3.9/site-packages/open3d/cpu/pybind.cpython-39-darwin.so (0x31d47ee48) and /Users/xxxxxxxx/anaconda3/envs/sphinx_env/lib/python3.9/site-packages/glfw/libglfw.3.dylib (0x339d45c48). One of the two will be used. Which one is undefined.
objc[16911]: Class GLFWWindow is implemented in both /Users/xxxxxxxx/anaconda3/envs/sphinx_env/lib/python3.9/site-packages/open3d/cpu/pybind.cpython-39-darwin.so (0x31d47eec0) and /Users/xxxxxxxxx/anaconda3/envs/sphinx_env/lib/python3.9/site-packages/glfw/libglfw.3.dylib (0x339d45cc0). One of the two will be used. Which one is undefined.
Warning: make sure gym is installed if you want to use the GymWrapper.
Total num of data item: 628, max_num_points: 9946
Creating all dataset with 20 demos
Total num of data item: 80, max_num_points: 9946
| Module                       |    #Params |      % |
|------------------------------+------------+--------|
| pos_rot_gripper_embed        |      1,536 |   0.01 |
| input_embed.weight           |      3,072 |   0.02 |
| input_embed.bias             |        512 |   0.00 |
| proprio_embed.weight         |      3,584 |   0.02 |
| proprio_embed.bias           |        512 |   0.00 |
| layers.0.ln1.weight          |        512 |   0.00 |
| layers.0.ln1.bias            |        512 |   0.00 |
| layers.0.mha.qkv_proj.weight |    786,432 |   4.15 |
| layers.0.mha.qkv_proj.bias   |      1,536 |   0.01 |
| layers.0.mha.out_proj.weight |    262,144 |   1.38 |
| layers.0.mha.out_proj.bias   |        512 |   0.00 |
| layers.0.ln2.weight          |        512 |   0.00 |
| layers.0.ln2.bias            |        512 |   0.00 |
| layers.0.linear1.weight      |  1,048,576 |   5.54 |
| layers.0.linear1.bias        |      2,048 |   0.01 |
| layers.0.linear2.weight      |  1,048,576 |   5.54 |
| layers.0.linear2.bias        |        512 |   0.00 |
| layers.1.ln1.weight          |        512 |   0.00 |
| layers.1.ln1.bias            |        512 |   0.00 |
| layers.1.mha.qkv_proj.weight |    786,432 |   4.15 |
| layers.1.mha.qkv_proj.bias   |      1,536 |   0.01 |
| layers.1.mha.out_proj.weight |    262,144 |   1.38 |
| layers.1.mha.out_proj.bias   |        512 |   0.00 |
| layers.1.ln2.weight          |        512 |   0.00 |
| layers.1.ln2.bias            |        512 |   0.00 |
| layers.1.linear1.weight      |  1,048,576 |   5.54 |
| layers.1.linear1.bias        |      2,048 |   0.01 |
| layers.1.linear2.weight      |  1,048,576 |   5.54 |
| layers.1.linear2.bias        |        512 |   0.00 |
| layers.2.ln1.weight          |        512 |   0.00 |
| layers.2.ln1.bias            |        512 |   0.00 |
| layers.2.mha.qkv_proj.weight |    786,432 |   4.15 |
| layers.2.mha.qkv_proj.bias   |      1,536 |   0.01 |
| layers.2.mha.out_proj.weight |    262,144 |   1.38 |
| layers.2.mha.out_proj.bias   |        512 |   0.00 |
| layers.2.ln2.weight          |        512 |   0.00 |
| layers.2.ln2.bias            |        512 |   0.00 |
| layers.2.linear1.weight      |  1,048,576 |   5.54 |
| layers.2.linear1.bias        |      2,048 |   0.01 |
| layers.2.linear2.weight      |  1,048,576 |   5.54 |
| layers.2.linear2.bias        |        512 |   0.00 |
| layers.3.ln1.weight          |        512 |   0.00 |
| layers.3.ln1.bias            |        512 |   0.00 |
| layers.3.mha.qkv_proj.weight |    786,432 |   4.15 |
| layers.3.mha.qkv_proj.bias   |      1,536 |   0.01 |
| layers.3.mha.out_proj.weight |    262,144 |   1.38 |
| layers.3.mha.out_proj.bias   |        512 |   0.00 |
| layers.3.ln2.weight          |        512 |   0.00 |
| layers.3.ln2.bias            |        512 |   0.00 |
| layers.3.linear1.weight      |  1,048,576 |   5.54 |
| layers.3.linear1.bias        |      2,048 |   0.01 |
| layers.3.linear2.weight      |  1,048,576 |   5.54 |
| layers.3.linear2.bias        |        512 |   0.00 |
| layers.4.ln1.weight          |        512 |   0.00 |
| layers.4.ln1.bias            |        512 |   0.00 |
| layers.4.mha.qkv_proj.weight |    786,432 |   4.15 |
| layers.4.mha.qkv_proj.bias   |      1,536 |   0.01 |
| layers.4.mha.out_proj.weight |    262,144 |   1.38 |
| layers.4.mha.out_proj.bias   |        512 |   0.00 |
| layers.4.ln2.weight          |        512 |   0.00 |
| layers.4.ln2.bias            |        512 |   0.00 |
| layers.4.linear1.weight      |  1,048,576 |   5.54 |
| layers.4.linear1.bias        |      2,048 |   0.01 |
| layers.4.linear2.weight      |  1,048,576 |   5.54 |
| layers.4.linear2.bias        |        512 |   0.00 |
| layers.5.ln1.weight          |        512 |   0.00 |
| layers.5.ln1.bias            |        512 |   0.00 |
| layers.5.mha.qkv_proj.weight |    786,432 |   4.15 |
| layers.5.mha.qkv_proj.bias   |      1,536 |   0.01 |
| layers.5.mha.out_proj.weight |    262,144 |   1.38 |
| layers.5.mha.out_proj.bias   |        512 |   0.00 |
| layers.5.ln2.weight          |        512 |   0.00 |
| layers.5.ln2.bias            |        512 |   0.00 |
| layers.5.linear1.weight      |  1,048,576 |   5.54 |
| layers.5.linear1.bias        |      2,048 |   0.01 |
| layers.5.linear2.weight      |  1,048,576 |   5.54 |
| layers.5.linear2.bias        |        512 |   0.00 |
| final_ln.weight              |        512 |   0.00 |
| final_ln.bias                |        512 |   0.00 |
| points_output.weight         |      2,048 |   0.01 |
| points_output.bias           |          4 |   0.00 |
| pos_output.weight            |      1,536 |   0.01 |
| pos_output.bias              |          3 |   0.00 |
| rot_output.weight            |      1,536 |   0.01 |
| rot_output.bias              |          3 |   0.00 |
| gripper_output.weight        |        512 |   0.00 |
| gripper_output.bias          |          1 |   0.00 |
| mode_output.weight           |      1,536 |   0.01 |
| mode_output.bias             |          3 |   0.00 |
| Total                        | 18,931,726 | 100.00 |
Running on device: mps
Running on device: mps

[1]    16911 bus error  python scripts/train_waypoint.py --config_path cfgs/waypoint/can.yaml

(sphinx_env) ➜  sphinx git:(main) ✗ /Users/xxxxxxxxx/anaconda3/envs/sphinx_env/lib/python3.9/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 14 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '

<img width="1512" alt="Image" src="https://github.com/user-attachments/assets/29b6db47-cca3-4302-967c-84f151b22886" />

I wonder what the problem is and how to solve it and make the training possible on my macbook pro. 

Thank you so much!

**Comment by hengyuan-hu** [comment]
Hey,

Sorry I have not used MPS for any training purpose before. I won't even trust pytorch MPS implementation to be correct honestly.
